docs/CNDP_README.md

```diff
@@ -0,0 +1,371 @@
+$ ./scripts/docker_setup.sh
+```
+Note: The script [cndp_reset_upf.sh](../scripts/cndp_reset_upf.sh) needs to be executed once before running [docker_setup.sh](../scripts/docker_setup.sh). After that we can execute [docker_setup.sh](../scripts/docker_setup.sh) multiple times if required. The script [cndp_reset_upf.sh](../scripts/cndp_reset_upf.sh) uses dpdk-devbind script from DPDK. To get a copy of it, execute the following command from the UPF's root directory:
+
+```
+$ wget https://raw.githubusercontent.com/DPDK/dpdk/main/usertools/dpdk-devbind.py -O dpdk-devbind.py
+$ chmod +x dpdk-devbind.py
+```
+
+Insert rules into relevant PDR and FAR tables
+
+```
+$ docker exec bess-pfcpiface pfcpiface -config /conf/upf.json -simulate create
+```
+9) From browser, use localhost:8000 to view the UPF pipeline in GUI. If you are remotely connecting to system via ssh, you need to setup a tunnel with local port forwarding.
+
+10) To stop the containers run following command
+
+```
+./scripts/cndp_reset_upf.sh
+```
+
+### Step 4: Run DPDK packet generator in System 2
+
+Build UPF docker image in System 2. Note: If you are behind a proxy make sure to export/setenv http_proxy and https_proxy
+From the top level directory call:
+
+```
+$ make docker-build
+```
+From system 2, bind the two interfaces used by pktgen to DPDK (used to send n/w packets to access/core ). Also setup huge pages in the system.
+
+Modify [pktgen_cndp.bess](../conf/pktgen_cndp.bess) script as follows.
+
+1) Update source and destination interface mac addresses of the access and core interface - smac_access, smac_core, dmac_access, dmac_core. Here smac_xxx corresponds to mac address of NIC in system 2 where we run pktgen and dst_xxx corresponds to mac address of NIC in system 1 which runs UPF pipeline.
+
+2) Update worker core ids to use core id in NUMA node where NIC is attached. For example if NIC is attached to NUMA node 1, use worker core ids in NUMA node 1. For example, `workers=[22, 23, 24, 25]`
+
+3) Bind the NICs to DPDK and note the vfio device number in "/dev/vfio"
+
+From the top level directory call: (**Note**: Update below command to set `cpuset-cpus`range same as worker core ids in step 2 above and use the vfio device number from step 3)
+
+```
+docker run --name pktgen -td --restart unless-stopped \
+        --cpuset-cpus=22-25 --ulimit memlock=-1 --cap-add IPC_LOCK \
+        -v /dev/hugepages:/dev/hugepages -v "$PWD/conf":/opt/bess/bessctl/conf \
+        -v /lib/firmware/intel:/lib/firmware/intel \
+        --device=/dev/vfio/vfio --device=/dev/vfio/119 --device=/dev/vfio/120 \
+        upf-epc-bess:"$(<VERSION)" -grpc-url=0.0.0.0:10514
+
+docker exec -it pktgen ./bessctl run pktgen_cndp
+```
+
+We can monitor if pktgen is sending packets using the following command:
+
+```
+docker exec -it pktgen ./bessctl monitor tc
+```
+
+If we need to stop sending packets at some point use the following command:
+
+```
+docker exec -it pktgen ./bessctl daemon reset
+```
```

Comment:
I prefer to remove this part and use the pktgen that I already have. However, I still need to open the PR for that :-(